### ISSUE TITLE
Create a "fetch" plugin prototype

### ISSUE THREAD

**Issue by FFY00** [conda/conda#11600]
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

There are several use cases that are desirable to be customizable that could be tackled by implementing a "repo interaction" plugin. Some examples are, authentication, replacing/modifying the transport layer, or even interfacing with custom repo implementations.

The idea behind this is that conda would provide a plugin that allows to completely replace the code where it interacts with the remote repository.

This has some challenges, as the current codebase is not in the best shape to implement this. And there are several parts of the code that interact with the repo and need hooking into.

Additionally, ideally, it would be good to also provide the original implementation as a plugin, so that people can re-use parts of it if they just need minor modifications.

**Comment by FFY00** [comment]
Even though the plugin scope will be quite broad, we should gather all the requirements in this issue, so that we can make sure all of them are addressed.

cc @jezdez

**Comment by jezdez** [comment]
Part of the use cases are covered in https://github.com/conda-incubator/ceps/pull/14

**Comment by FFY00** [comment]
Okay, there are several code paths which this plugin can impact, depending on the design.

- `conda.connection.adapters`
	- Essentially exposes the `requests` package
- `conda.connection.session.CondaSession`
	- Abstracts `requests.Session`, which handles remote interaction (HTTP(S), FTP, etc.)
- `conda.subdir_data.fetch_repodata_remote_request`
	- Fetches the repo data from the remote repo
		- Receives a URL and the target repodata file and performs an HTTP request
- `conda.core.subdir_data.SubdirData._load`
	- Loads the
- `conda.trust.signature_verification._SignatureVerification._fetch_channel_signing_data`
	- Fetches signing data from the repo

Currently, the repo interaction is implemented by a couple functions which receive the URL and repodata name. They then use `CondaSession` to access the remote, and also implement cache. Changing this is tricky as the code is closely coupled. The channel loading by `SubdirData.query_all`, but I am not sure if there are hidden places that go around that code path.

After carefully reviewing our options, I think the best design for now is replacing the url that gets passed around with a callable that fetches the data for a given repodata name. However, this is not optimal, and in my opinion it'd be good to uncouple the repodata from these steps, passing just a callable that already has a repodata in mind, however that is not possible right now.

Regarding cache, I think optimally we should have a default cache implementation, but maybe eventually allow plugin authors to re-implement it for their own repos. But there are still some other questions up in the air, like the signing data, which are currently implemented with assumptions regarding the repo transport that will need to be handled.

Even though we have a good plan for how to integrate this plugin with the conda internals, I am still not sure exactly how the plugin declaration design should be. The easiest way would be for a plugin to inject new channels directly into conda, but I am not sure if that's the UX we want, and if it isn't, it is still a question to me of how channels would be declared to specify a specific plugin/scheme to be used.

**Comment by jezdez** [comment]
> Okay, there are several code paths which this plugin can impact, depending on the design.
> 
> * `conda.connection.adapters`
>   
>   * Essentially exposes the `requests` package
> * `conda.connection.session.CondaSession`
>   
>   * Abstracts `requests.Session`, which handles remote interaction (HTTP(S), FTP, etc.)
> * `conda.subdir_data.fetch_repodata_remote_request`
>   
>   * Fetches the repo data from the remote repo
>     
>     * Receives a URL and the target repodata file and performs an HTTP request
> * `conda.core.subdir_data.SubdirData._load`
>   
>   * Loads the
> * `conda.trust.signature_verification._SignatureVerification._fetch_channel_signing_data`
>   
>   * Fetches signing data from the repo
> 
> Currently, the repo interaction is implemented by a couple functions which receive the URL and repodata name. They then use `CondaSession` to access the remote, and also implement cache. Changing this is tricky as the code is closely coupled. The channel loading by `SubdirData.query_all`, but I am not sure if there are hidden places that go around that code path.

I think this would imply that some refactoring would be required to have a common API to load files from repos, e.g. a "CondaLoader" class or something, whose methods could be overridden by plugins. That way we'd be able to provide a number of plugin hooks for the various pieces of the network interaction, without limiting future expanding of the API.

> After carefully reviewing our options, I think the best design for now is replacing the url that gets passed around with a callable that fetches the data for a given repodata name. However, this is not optimal, and in my opinion it'd be good to uncouple the repodata from these steps, passing just a callable that already has a repodata in mind, however that is not possible right now.

Could we pass around an instance of a `CondaLoader` class to do the same, so we get integration into the plugin system? 

> Regarding cache, I think optimally we should have a default cache implementation, but maybe eventually allow plugin authors to re-implement it for their own repos. But there are still some other questions up in the air, like the signing data, which are currently implemented with assumptions regarding the repo transport that will need to be handled.

Agreed the cache implementation would be a really excellent plugin hook, and could be something to look into in the future. Meanwhile, it could just be a method implemented in the `CondaLoader` class to be turned into a plugin hook later. 

> Even though we have a good plan for how to integrate this plugin with the conda internals, I am still not sure exactly how the plugin declaration design should be. The easiest way would be for a plugin to inject new channels directly into conda, but I am not sure if that's the UX we want, and if it isn't, it is still a question to me of how channels would be declared to specify a specific plugin/scheme to be used.

Listing out the calls to fetch data from repos would be a good step, to see what we would need to cover by the plugin API and see opportunities for refactoring. The current code is not extensible so and not part of the "public API", so refactoring is technically possible.

**Comment by FFY00** [comment]
I will reply to you later, but I should document the next steps.

- To have something tangible to point to, we should have a plugin prototype that hooks into `SubdirData`
- We should decide on the UX, specifically how channels that use a repo plugin are handled

**Comment by dholth** [comment]
@FFY00 I am interested in this. We want to do incremental repodata and we want to try https://github.com/conda/conda/issues/11608

**Comment by FFY00** [comment]
Humm, interesting. I wasn't planning on it but it makes sense. We can make the fetching functions optionally async, enabling those use-cases.

I am currently working on the prototype that just hooks into `SubdirData`.

**Comment by dholth** [comment]
It will be interesting to figure out how to make things look "not necessarily requests-like"

**Comment by FFY00** [comment]
Just a quick update. I have opened a draft PR, #11788, with a prototype. The goal is to provide a design for the plugin data structures and extract the current implementation to what will become the default implementation plugin.

After the prototype is working, the following steps will be:
  - Migrate the tests to the new structure (like we did for the solver plugin)
  - Open a PR against the `plugins` branch implementing the plugin hook (and use the work done in the prototype as the default plugin)
  	- This is currently blocked due to us not knowing yet how we want to design the UX

**Comment by travishathaway** [comment]
@FFY00, I've updated the title of this issue to, "Create a "fetch" plugin prototype". This was done in order to create a large epic to organize this work. You can find that here: https://github.com/conda/conda/issues/11821 I did this because it will make it easier for us to organize all the work around this feature.

Please let me if this okay. I will move the important bits from this issue to that one so we don't loose any details.

**Comment by FFY00** [comment]
The repodata fetching in the prototype is now working, the next step in the package fetching.

**Comment by FFY00** [comment]
The package fetching refactor conflicts with #11841, I think the best way to handle it is to work on top of that work, as it is my understanding that should be merged soon-ish.

**Comment by FFY00** [comment]
Currently, we are blocked on deciding the UX for the plugin, what happens when users add a new channel which needs this plugin specifically.

My proposal would be when a user adds a channel, they can pass an argument specifying the "type" of channel it is (eg. `conda-simple`, `special-flavor`, etc.).

```
$ conda config --add channels https://my-project.com/conda --type special-flavor
```

Each "fetch" plugin would register all the channel types they support in the metadata. This would enable Conda to find the plugin required to handle the channel, and would even allow alternative implementations for each type to be provided (eg. a plugin that uses [powerloader](https://github.com/mamba-org/powerloader) as an alternative to the default Conda "fetch" handler), promoting innovation and giving power back to the community.

```
$ conda config --add channels https://my-project.com/conda --type special-flavor
This channel requires a Conda plugin. There are 2 plugins that implement support for this channel type, please select one to install:
  1) special-plugin
     Description: Plugin to interact with Special(TM) channels
     Author: Mega Corporation
  2) special-plugin-multithreaded
     Description: Multithreaded plugin to interact with Special(TM) channels
     Author: Your Friendly Neighbourhood OSS Maintainer
```

---

This is just a proposal, but while considering alternatives or tweaks to this scheme, it is **critical** to not add any dependencies on the transport layer, as the whole point of this plugin is to abstract that away.

---

cc @jezdez @beeankha @dholth @travishathaway

**Comment by travishathaway** [comment]
Hi @FFY00,

This is an interesting proposal. I am now wondering how this would be laid out in our configuration files (i.e. `.condarc`). Here's my initial thought:

```yaml
channels:
  - defaults
  - https://package-corp.com:
     type: basic-auth
```

This config layout would be backwards compatible with existing user configs and allow us to add new information about each channel like you suggested.

I would also think that it would be nice for users to be able to list the available channel types if they ever needed to (as you install new plugins, new channel types would be made available). For example, we could put it in the output of `conda info`. Perhaps we could even add a special `--plugins` option that would only show plugin specific environment settings.

Also regarding channel types, I think this should be a new hook that we expose for plugin authors. Our current work on #11826 could help inform this.
